### PR TITLE
Optimize MutableDenseHashTable's memory usage and performance in case of large value_shape

### PR DESCRIPTION
**Pull request by evanrang**
When value_shape is very large, we'll store a pointer in key_buckets element, and store table values into a vector of value_buckets, which containers no empty values. This optimize can decrease 15% memory usage in a test case of value_dtype = int64 and value_shape = [16], also get better lookup/insert performance.
If value_shape is very small or key/value dose not support memcpy, we'll back to the version of single key_buckets and value_buckets_.

### CODE REVIEW

**Comment by rangjiaheng** [comment]
@rohan100jain hi, any suggestions ?

**Comment by gbaned** [comment]
@rangjiaheng  Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@rangjiaheng  Any update on this PR? Please. Thanks!

File: tensorflow/core/kernels/lookup_table_op.cc [hunk 1]
```diff
@@ -792,10 +1020,18 @@ class MutableDenseHashTable final : public LookupInterface {
...
   PersistentTensor empty_key_;
   uint64 empty_key_hash_;
   PersistentTensor deleted_key_;
   uint64 deleted_key_hash_;
+
+  int32 key_bytes_;
+  int32 value_bytes_;
+  bool use_value_ptr_;
+
+  int64 value_index_; TF_GUARDED_BY(mu_)
```
Comment (rohan100jain) [inline tensorflow/core/kernels/lookup_table_op.cc:1033]: Add comments describing what these member variables are for

File: tensorflow/core/kernels/lookup_table_op.cc [hunk 2]
```diff
@@ -391,10 +391,25 @@ class MutableDenseHashTable final : public LookupInterface {
           errors::InvalidArgument("Empty and deleted keys cannot be equal"));
     }
 
+    key_bytes_ = key_shape_.num_elements() * sizeof(K);
+    value_bytes_ = value_shape_.num_elements() * sizeof(V);
+    if (DataTypeCanUseMemcpy(key_dtype()) &&
+        DataTypeCanUseMemcpy(value_dtype()) &&
+        sizeof(K) <= sizeof(void*) && value_bytes_ > sizeof(void*)) {
```
Comment (rohan100jain) [inline tensorflow/core/kernels/lookup_table_op.cc:398]: The constraint of sizeok(K) <= sizeof(void*) is unclear to me....

File: tensorflow/core/kernels/lookup_table_op.cc [hunk 3]
```diff
@@ -429,6 +444,7 @@ class MutableDenseHashTable final : public LookupInterface {
         deleted_key_.AccessTensor(ctx)->template shaped<K, 2>({1, key_size});
     const int64 bit_mask = num_buckets_ - 1;
     // TODO(andreasst): parallelize using work_sharder
+    const auto* default_flat_ptr = default_flat.data();
```
Comment (rohan100jain) [inline tensorflow/core/kernels/lookup_table_op.cc:447]: nit: Move this to line 435

File: tensorflow/core/kernels/lookup_table_op.cc [hunk 4]
```diff
@@ -723,9 +852,10 @@ class MutableDenseHashTable final : public LookupInterface {
     num_entries_ = 0;
 
     const int64 key_size = key_shape_.num_elements();
+    const int64 allocate_size = key_bytes_ / sizeof(K);
     Tensor* key_buckets_tensor;
     TF_RETURN_IF_ERROR(ctx->allocate_persistent(
-        key_dtype(), TensorShape({num_buckets_, key_size}), &key_buckets_,
+        key_dtype(), TensorShape({num_buckets_, allocate_size}), &key_buckets_,
```
Comment (rohan100jain) [inline tensorflow/core/kernels/lookup_table_op.cc:858]: Its unclear to me that when use_value_ptr_ is false also we are modifying key_buckets_?

File: tensorflow/core/kernels/lookup_table_op.cc [hunk 5]
```diff
@@ -626,23 +730,48 @@ class MutableDenseHashTable final : public LookupInterface {
...
-                SubtleMustCopyIfIntegral(value_matrix(i, j));
+          if (use_value_ptr_) {
+            const auto* src = (const char*) (key_matrix.data() + i * key_size);
+            auto* dst =
+                (char*) key_buckets_matrix.data() + bucket_index * key_bytes_;
+            memcpy(dst, src, key_size * sizeof(K));      // key element
+
+            src = (const char*) (value_matrix.data() + i * value_size);
+            dst += key_size * sizeof(K);
+            if (*(void**) dst == nullptr) {              // reuse value chunk
```
Comment (rohan100jain) [inline tensorflow/core/kernels/lookup_table_op.cc:758]: So looks like we keep adding to the value_buckets_vec_ but do we ever reclaim that memory? Or does it keep on increasing? Would that cause adverse performance on workloads that have a lot of Inserts and Deletes?

**Review by rohan100jain — CHANGES_REQUESTED** [review]
Thanks a lot for the contribution! Have you done some benchmarking on various workloads to see what the impact is?

**Comment by gbaned** [comment]
@rangjiaheng Can you please check @rohan100jain's comments and resolve conflicts?. Thanks!

**Comment by gbaned** [comment]
@rangjiaheng Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 19 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**